Multiple rooms! A variant of the Sessions idea.

chadoh opened the pull request:
Fixes #7
## What is it

The index page now has a description of what Lean Coffee is, and a list of different rooms. You can give rooms a custom url. I'd imagine there will be long-lived rooms used by different teams, such as `/engineering` and `/advertising`. People from different teams will probably end up with their browser autocompleting their room URL, and, most of the time, they won't even have to think about the fact that other rooms exist.

Each room gets a view identical to what the app had before, except there's also a footer with some links.

Once topics are in the "talked about" column, an "archive" action button will be available on them. The same archive button also appears at the top of the "talked about" column, and allows bulk-archiving all "talked about" topics.

Archived topics can be viewed in a "room archive" page. They can also be searched, using plain ol' Rails searching.
## How it looks

![rooms list](https://cloud.githubusercontent.com/assets/221614/8398327/cbff47a4-1db7-11e5-87ad-4bfb492798ef.png)

---

![edit page](https://cloud.githubusercontent.com/assets/221614/8394020/2616e4b8-1cf5-11e5-8607-97d1a745598b.png)

---

![room](https://cloud.githubusercontent.com/assets/221614/8399889/5dd5bbdc-1de4-11e5-9b7b-e5d3f559c495.png)

---

![archive](https://cloud.githubusercontent.com/assets/221614/8399894/678fb6e6-1de4-11e5-8ed2-e4f6692d4a18.png)

chadoh committed:
Merge branch 'remove-default-readme'

* remove-default-readme:
  Remove the rails-generated README.rdoc

chadoh committed:
Merge branch 'update-bundle'

* update-bundle:
  Update readme, too

chadoh committed:
Merge branch 'update-bundle'

* update-bundle:
  Get it to run; update gems and readme and db names

chadoh committed:
Add scaffolding for rooms

chadoh committed:
Remove scaffolding kruft; migrate

chadoh committed:
Make the URLs for rooms behave as desired

`/room-name` now routes to that room, and Rails generates all of the
URLs correctly.

chadoh committed:
Add validations and indexes

chadoh committed:
Merge branch 'master' into rooms

* master:
  Remove sessions stuff from schema file

chadoh committed:
Scope topics by room; It works!

With these changes, it works! Peeps can now create multiple rooms, with
friendly urls, and all of their stuff will be kept separate.

The index page is far from beautiful, right now. Beautification comes
next!

chadoh commented:
Oh, ha. I just saw that @kpilka submitted a different approach to this problem, #15, a day after @jgorset asked me to look into it. :smile: 

Oh well!

If you don't mind, I'm going to keep going with my approach. What I like about it:
- The "board" pages, with all of the topics, stay as simple and clutter-free as before. I love the current clutter-free feel of this app, and I want to keep it.
- Static URLs for different rooms/teams feels like better long-term UX. I don't want to have to go to the index every time I want to start a new LeanCoffee session. And I don't want to have to have a teammate send me the URL of the current session, or go digging through a UI to find the correct session.
- "Archives", grouped by `created_at` date (which will 99% of the time be when they were discussed, too), seem like a simpler and more user-friendly approach to retaining historic data.

Let me know what y'all think!

chadoh committed:
Style and add content to home page

The home page looks pretty now!

chadoh committed:
Style the form pages

chadoh commented:
Styling that up was a fun way to spend a Saturday afternoon! I attached screenshots. I think you'll love it. :smile:

chadoh committed:
Add footer to topics board pages

When you're in a room, you might want to edit its details or get back to
the home page without manually editing the URL. Now you can! Just scroll
down.

I really liked that the page by default only shows the board, and that
as soon as you start to scroll down, you see the footer. I copied this
on the home page, to make the list of rooms take up the entire view
height, as well. And since people who already use the app don't need to
see the intro stuff, I made the main links back to the homepage go to
just that bottom section, by default.

Getting the text in the footer to be vertically centered was totally
crazy-bonkers-crazy! Wow. CSS!! Even with flex-box, it's still totally
whacked sometimes. Although flexbox certainly made it a lot better!
After learning some flexbox tricks, I went and made the grid of rooms on
the homepage look nicer, too.

The CSS feels sorta unorganized to me. I've never worked on a team where
it actually feels organized. I feel like this project is reaching a
point where a better organizational scheme would be really useful. But I
don't know if it's worth learning how to do that. I feel like knowing
about how to organize CSS files is a designer's job. Maybe that's lazy
of me. :-/

chadoh committed:
Allow archiving topics

The "archive" button only shows up if the topic has been talked about.

For now, one can only archive topics one at a time.

chadoh committed:
Allow archiving entire Talked About column at once

I also cleaned up the CSS a little bit, extracting the "hover-options"
stuff into a common class that now lives in application.sass.

chadoh committed:
Add searchable archives

chadoh commented:
@jgorset it's ready! Please feel free to give very honest feedback on everything. I'll happily add in anything y'all feel is missing. I think it looks pretty and works well!

jgorset commented:
Thanks @chadoh, and I will! Prepare your inbox!

jgorset commented:
I like how that panel beneath the notes is always just out of sight regardless of your window's size. Neat.

jgorset commented on .gitignore:
```diff
@@ -15,3 +15,4 @@
 /log/*
 !/log/.keep
 /tmp
+.DS_Store
```
This is a pretty common thing to do, but after a while local ignore files get pretty cluttered if lots of different people with lots of different operating systems and editors work on it. As a rule, we think local ignore files should only contain files that pertain to the repository itself.

That's not to say we should version control `.DS_Store`! You can put that in your [global ignore file](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore), though, which we think make more sense because really it's about your system and not the repository.

chadoh replied:
I like that policy. Line removed!

jgorset commented on app/assets/javascripts/angular/controllers/topics_ctrl.coffee:
```diff
@@ -15,6 +15,15 @@ angular.module "lean-coffee"
+    @archiveAll = =>
+      if confirm('Are you sure you want to archive all discussed topics?')
+        topicsService.archiveAll()
+        topicsService.reload()
```
How come `archiveAll` needs to reload, but `archive` doesn't?

chadoh replied:
`archive` works on one topic. Like `update` and `delete`, once that topic is updated, Angular (somehow) knows to update the data it's holding onto, and its two-way data binding causes the card for that topic to get refreshed (in the case of `delete`, it causes it to be removed).

That "somehow" is a mystery I didn't fully probe. I'm not sure if Angular listens for the return value from the server, or if it gets the message from Pusher. Maybe whichever one is first. It's possible that all `TopicsController` json responses could be empty.

In any case, [`topicsService.topics` is being fetched at page load and then never updated](https://github.com/chadoh/lean-coffee/blob/rooms/app/assets/javascripts/angular/services/topics_service.coffee#L3). 

You know, now that I'm thinking about this, I don't think `archiveAll` is going to get pushed to all clients correctly. I am in fact [publishing about each one that was updated](https://github.com/chadoh/lean-coffee/blob/rooms/app/controllers/topics_controller.rb#L47), but [the Angular template I wrote](https://github.com/chadoh/lean-coffee/blob/rooms/app/views/rooms/show.html.slim#L27) wasn't dealing with it properly and hiding them, and my crappy solution at the time was just to reload all topics. But that will only work for the browser making the request. I shouldn't need this `reload`. I'll improve this now.

Thanks for asking about it!

jgorset replied:
Aha! I thought I was just missing something, but I'm glad that I could be of some assistance!

To be fair I have probed it no more than you, and so I'd also be curious to know whether Angular updates its model with the response from the server. I'm inclined to think it wouldn't, though, as you'd have to return the same representation from the server on every request and that seems a lot for it to presume.

jgorset replied:
Either way, I'm interested to see how you tackle this.

chadoh replied:
When a browser is the one that makes an ajax call, [this `socket_id` setting](https://github.com/chadoh/lean-coffee/blob/rooms/app/assets/javascripts/application.coffee#L32) causes it to ignore the resulting Pusher update. It itself is updated the traditional way, using the payload of the ajax request.

At least, that's how it works now. I'm experimenting with removing the `socket_id` so that all of the pages are updated the same way. But right now, that causes nasty infinite loops of ajax requests! 

For some reason, on page load, an `update` request is fired for _every single topic on the page_. **_FOUR TIMES**_. If you have two topics on the page, Angular sends off 8 ajax requests right after page load! No idea why yet. But if I naively remove this `socket_id`, then each of those kicks off another (which is also mysterious), and :boom:

jgorset replied:
Oh no! Our bad! I can't think of any reason why it would do that, though – really it shouldn't need to update anything at all when the page loads.

chadoh replied:
No worries! I tracked it down and fixed it all up in 3876d46. Which also allowed removing the `reload` call here. I think both I and the code are better now because of it! :smile:

jgorset commented on app/assets/stylesheets/application.sass:
```diff
@@ -9,33 +9,42 @@
-  background: rgba(245, 125, 45, 0.9)
-  color: #fff
+  background: $primary-color
+  color: $light-color
```
Thanks for refactoring this!

chadoh replied:
My pleasure! :nail_care:

jgorset commented on app/assets/stylesheets/application.sass:
```diff
@@ -48,3 +57,87 @@ p
+    margin: 3rem 0
+    @media(min-width: 768px)
+      +flex(1)
+      font-size: 6vw
```
Woa woa! I didn't know about viewport-sized typography, but today I learned! Thanks!

chadoh replied:
I didn't know about `vh` and `vw` until I saw them in this project! I got pretty into those units. :stuck_out_tongue_closed_eyes:

jgorset replied:
Wait – they were already here? Then I should perhaps peruse my pull requests as closely as this one more often. :wink:

chadoh replied:
`vh` was used to calculate the height of the columns and the page. I went crazy and used `vh` and `vw` for sizing text, too!

jgorset commented on app/assets/stylesheets/application.sass:
```diff
@@ -48,3 +57,87 @@ p
+            border-style: solid
+          div
+            @media(min-width: 768px)
+              display: table // vertical alignment is insane
```
Yes. Yes it is.

jgorset commented on app/controllers/rooms_controller.rb:
```diff
@@ -0,0 +1,69 @@
+
+  def show
+    @room = Room.find_by(slug: params[:slug])
+    raise ActiveRecord::RecordNotFound, "Couldn't find a room for that URL" unless @room
```
If you use `find_by!`, ActiveRecord willl raise this exception for you. :wink:

chadoh replied:
Perfect! Thanks for that. Updated.

jgorset commented on app/controllers/rooms_controller.rb:
```diff
@@ -0,0 +1,69 @@
+  def archive
+    topics = @room.topics.order('created_at DESC')
+    topics.where!('title ilike ?', "%#{params[:q]}%") if params[:q]
+    @grouped_topics = topics.group_by{ |t| t.created_at.to_date }
```
Given its name and location (`rooms#archive`), I assumed this method would archive the room. That's not what it does, though, but rather it's a searchable archive of topics in the room. Which is neat. But it should probably be a member action on the topics controller instead. The route could be `/name-of-room/topics/archived?q=title` or something. What do you think?

chadoh replied:
I was originally going to do it exactly like you suggest!

But then the room's title ended up being "Awesome Room Archive". And I had gotten really enamored with the super clean user-facing URLs in this app. So a GET request to a URL of `/awesome/archive` seemed like it made sense!

From an end-user perspective, I think my approach does make sense—I'm GETting an archive, not PUTting something in an archive. But I can definitely see why this would be confusing, from a code readability perspective.

I just thought of maybe a simpler change that could keep the clean URL and communicate better with other code-readers. What if I rename it `archives` instead of `archive`?

chadoh replied:
Since renaming it to `archives` was such an easy change, I went ahead and did it already. Try it out, and let me know if you still think it'll be confusing for future devs and not worth the (probably negligible) user nicety of a cleaner URL.

jgorset replied:
Good point; from the perspective of the user, a HTTP GET to `/room/archive` can only mean one thing and it's shorter than `/room/topics/archive`. I agree with this (and especially since you renamed it to `archives`).

jgorset replied:
![Kim approved](http://i.imgur.com/ybyfarR.gif)

jgorset commented on app/controllers/topics_controller.rb:
```diff
@@ -90,6 +110,10 @@ def destroy
 
   private
 
+  def get_room
```
I'd name this one `set_room`, too, for consistency with `set_room` on your rooms controller.

chadoh replied:
I didn't even notice that inconsistency. Thanks for pointing it out! Fixed.

chadoh committed:
Remove .DS_Store from local .gitignore

I've started ignoring them globally.

chadoh committed:
Prefer `find_by!` rather than `find_by`

chadoh committed:
Rename get_room to set_room in TopicsController

chadoh committed:
Merge branch 'master' into rooms

* master:
  Use POST for voting

chadoh committed:
Rename "archive" to "archives"

This sounds like the more common way to describe such a thing, and
communicates better when reading the controller that it's not an action
that puts the entire room in an archive.

chadoh commented:
> I like how that panel beneath the notes is always just out of sight regardless of your window's size. Neat.

That was all you guys! I put a footer down there, and the `vh` styles you had already added made it do that. I liked it so much, and it was such an easy thing to do, that I added the same effect to the `#room-list` div on the home page. :smile:

chadoh committed:
Always use WebSockets, no traditional ajax

This fixes the "archive all" button and makes it work for all connected
clients.

It also resolves an issue I had noticed: on page load, Angular fired an
"update" request four times for every topic on the page. This seems to
have been caused by this:

    ng-submit="topicsCtrl.update(topic); topicCtrl.edit_topic = false"

Apparently this caused Angular to call `topicsCtrl.update` every single
render pass!

Once again, Angular's janky fake JS that it squishes unnaturally into
HTML makes me long for the simplicity of React's all-JS, just-JS
approach. HTML is a simple, predictable, declarative language. It makes
way more sense to put it into your JS than to force bastardized JS into
HTML.

Anyhow. I stopped Angular from POSTing every time it renders.

Note that it was also rendering every time it got a Pusher update.

If you're paying close attention, you'll notice the loop:

angular renders page ->
  sends ajax update ->
    server publishes update to Pusher ->
      angular, subscribed to Pusher channel, re-renders ->
        :boom:

The reason this didn't cause infinite loops previously is the nice
`socket_id` trick. The Pusher JS client had a `socket_id` set, which
Angular sent along to the server with the payload, and when the
controller published to Pusher, it included it. The Pusher client would
then see this `socket_id`, and say "oh, I initiated that, I don't need
to do anything."

So, having fixed the update-on-render bug, I was able to remove the
`socket_id` crutch. This then allowed removing all traditional ajax
responses, since the page sending the request is now updated via
Pusher/WebSockets, just like any other subscribed browser.

chadoh commented:
@jgorset no problem! Github told me that it had conflicts. :smile: 

I believe I addressed all of the issues, large and small, that you noticed. This might be ready for merge!

jgorset commented:
Thank you @chadoh! I believe it is! We look forward to having rooms in our next lean coffee.

mathiasbno commented:
This is great :+1:

ohenrik commented:
:cake: :D

timkurvers commented:
Fantastic stuff :+1:
